src/blk/pmem/PMEMDevice.cc

```diff
@@ -54,10 +59,61 @@ int PMEMDevice::_lock()
   return 0;
 }
 
+static int pmem_check_file_type(int fd, const char *pmem_file, uint64_t *total_size)
+{
+  int rc = 0;
+  struct stat file_stat;
+  char spath[PATH_MAX];
+  char npath[PATH_MAX];
+  char *real_path, *base_name;
+  FILE *sfile;
+
+  rc = ::fstat(fd, &file_stat);
+  if (rc) {
+    return -1;
+  }
+
+  if ((file_stat.st_mode & S_IFCHR) != S_IFCHR) {
+    return -1;
+  }
+
+  snprintf(spath, PATH_MAX, "/sys/dev/char/%d:%d/subsystem",
+           major(file_stat.st_rdev), minor(file_stat.st_rdev));
+
+  real_path = realpath(spath, npath);
```

Comment:
```c++
namespace fs = std::filesystem
if (!fs::is_character_file(pmem_file)) {
  return -EINVAL;
}
struct stat file_stat;
if (::fstat(fd, &file_stat)) {
  return -EINVAL;
}
auto char_dir = fmt::format("/sys/dev/char/{}:{}",
                            major(file_stat.st_rdev), 
                            minor(file_stat.st_rdev));
if (auto subsys_path = char_path / "subsystem";
    fs::read_symlink(subsys_path).filename().string() != "dax") {
  return -EINVAL;
}
if (total_size == nullptr) {
  return 0;
}
if (std::ifstream size_file(char_path / "size"); size_file)
  size_file >> *total_size;
  return size_file ? 0 : -EINVAL;
} else {
  return -EINVAL;
}
```

i'd suggest rewrite this part usinf `std::filesystem()` for better readability.

Reply:
Can we revise it later?

Reply:
sure.

Reply:
#46122